File: lightroom_tagger/core/retry.py
"""Configurable retry with exponential backoff for provider calls."""

from __future__ import annotations

import time
from typing import Any, Callable, TypeVar

from lightroom_tagger.core import cancel_scope
from lightroom_tagger.core.provider_errors import (
    NOT_RETRYABLE_ERRORS,
    RETRYABLE_ERRORS,
    ProviderError,
)

T = TypeVar("T")


class CancelledRetryError(Exception):
    """Raised by :func:`retry_with_backoff` when ``cancel_check`` returns truthy
    during a backoff sleep. Callers should treat it as an orderly cancel rather
    than a provider failure (no retry, no fallback).
    """


def _interruptible_sleep(
    total_seconds: float,
    cancel_check: Callable[[], bool] | None,
    step: float = 0.5,
) -> None:
    """Sleep ``total_seconds`` in small slices, aborting early on cancel.

    Without this, ``time.sleep(32)`` inside a retry backoff holds the worker
    for the full duration regardless of ``runner.is_cancelled`` — that is the
    exact failure mode that left job ``b141dbcc`` CPU-pegged for 9 minutes
    after a cancel.
    """
    if cancel_check is None:
        time.sleep(total_seconds)
        return
    remaining = max(0.0, float(total_seconds))
    while remaining > 0:
        if cancel_check():
            raise CancelledRetryError("cancel requested during backoff")
        slice_ = step if remaining > step else remaining
        time.sleep(slice_)
        remaining -= slice_


def retry_with_backoff(
    fn: Callable[[], T],
    retry_config: dict[str, Any],
    log_callback: Callable[[str, str], None] | None = None,
    cancel_check: Callable[[], bool] | None = None,
) -> T:
    """Call *fn* with retry on retryable ``ProviderError`` subclasses.

    Parameters
    ----------
    fn:
        Zero-arg callable (bind args with ``functools.partial`` or lambda).
    retry_config:
        ``{"max_retries": int, "backoff_seconds": list[float], "respect_retry_after": bool}``
    log_callback:
        Optional ``(level, message)`` logger.

    Returns the result of *fn* on success.
    Raises the last ``ProviderError`` if all retries are exhausted.
    Raises immediately for non-retryable errors.
    """
    max_retries: int = retry_config.get("max_retries", 3)
    backoff: list[float] = retry_config.get("backoff_seconds", [2, 8, 32])
    respect_retry_after: bool = retry_config.get("respect_retry_after", True)

    # Only opt into the thread-local scope when the current thread has one
    # installed. Otherwise we keep the pre-cancel behaviour exactly —
    # ``time.sleep(wait)`` called once per attempt — so existing tests and
    # timing assumptions stay untouched.
    if cancel_check is None and cancel_scope.has_active_scope():
        cancel_check = cancel_scope.is_cancelled

    last_error: ProviderError | None = None

    for attempt in range(1 + max_retries):
        # Short-circuit before each attempt so a cancel requested while the
        # previous call was in-flight (or during the next backoff) prevents
        # both another call and another sleep.
        if cancel_check is not None and cancel_check():
            raise CancelledRetryError("cancel requested before retry attempt")

        try:
            return fn()
        except tuple(NOT_RETRYABLE_ERRORS) as exc:
            raise
        except tuple(RETRYABLE_ERRORS) as exc:
            last_error = exc  # type: ignore[assignment]

            if attempt >= max_retries:
                break

            wait = backoff[attempt] if attempt < len(backoff) else backoff[-1]

            if (
                respect_retry_after
                and isinstance(exc, ProviderError)
                and getattr(exc, "retry_after", None) is not None
            ):
                wait = exc.retry_after  # type: ignore[assignment]

            if log_callback:
                log_callback(
                    "warning",
                    f"Retry {attempt + 1}/{max_retries} after {type(exc).__name__}: "
                    f"{exc} — waiting {wait}s",
                )

            _interruptible_sleep(wait, cancel_check)

    raise last_error  # type: ignore[misc]

File: apps/visualizer/backend-ts/src/providers/retry.ts
/**
 * Configurable retry with exponential backoff for provider calls.
 * Port of `core/retry.py`.
 *
 * The Python version needed `_interruptible_sleep` and a thread-local
 * `cancel_scope`, because `time.sleep(32)` inside a backoff held the worker for
 * the full duration regardless of `runner.is_cancelled` — the exact failure that
 * left job `b141dbcc` CPU-pegged for nine minutes after a cancel. And threading a
 * cancel check through runner → handler → description_service → analyzer →
 * dispatcher → retry was five layers of signature plumbing, hence the
 * thread-local.
 *
 * Neither is needed here. `await sleep()` does not hold anything, so the check
 * before each slice is only about *promptness*, and an explicit optional
 * `cancelCheck` threads through async calls without ceremony. The thread-local
 * scope has no analogue and is not reproduced.
 */
import { isNotRetryableError, isRetryableError, ProviderError } from './errors.js';

export type LogCallback = ((level: string, message: string) => void) | null;
export type CancelCheck = (() => boolean) | null;

/**
 * A cancel was requested during a backoff.
 *
 * Callers must treat it as an orderly stop rather than a provider failure: no
 * retry, no fallback.
 */
export class CancelledRetryError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'CancelledRetryError';
  }
}

export interface RetryConfig {
  max_retries?: number;
  backoff_seconds?: number[];
  respect_retry_after?: boolean;
}

/** Sleep in slices so a cancel is noticed within ~500ms rather than ~32s. */
async function interruptibleSleep(
  totalSeconds: number,
  cancelCheck: CancelCheck,
  stepSeconds = 0.5,
): Promise<void> {
  const sleep = (s: number): Promise<void> =>
    new Promise((resolve) => {
      setTimeout(resolve, s * 1000);
    });

  if (cancelCheck === null) {
    await sleep(totalSeconds);
    return;
  }
  let remaining = Math.max(0, totalSeconds);
  while (remaining > 0) {
    if (cancelCheck()) throw new CancelledRetryError('cancel requested during backoff');
    const slice = remaining > stepSeconds ? stepSeconds : remaining;
    await sleep(slice);
    remaining -= slice;
  }
}

/**
 * Call `fn` with retry on retryable provider errors.
 *
 * Returns `fn`'s result, or throws the last retryable error once the ladder is
 * exhausted. A non-retryable error is rethrown immediately.
 */
export async function retryWithBackoff<T>(
  fn: () => Promise<T>,
  retryConfig: RetryConfig,
  opts: { logCallback?: LogCallback; cancelCheck?: CancelCheck } = {},
): Promise<T> {
  const maxRetries = retryConfig.max_retries ?? 3;
  const backoff = retryConfig.backoff_seconds ?? [2, 8, 32];
  const respectRetryAfter = retryConfig.respect_retry_after ?? true;
  const logCallback = opts.logCallback ?? null;
  const cancelCheck = opts.cancelCheck ?? null;

  let lastError: unknown = null;

  for (let attempt = 0; attempt <= maxRetries; attempt += 1) {
    // Checked before each attempt so a cancel requested while the previous call
    // was in flight prevents both another call and another sleep.
    if (cancelCheck !== null && cancelCheck()) {
      throw new CancelledRetryError('cancel requested before retry attempt');
    }

    try {
      return await fn();
    } catch (e) {
      if (isNotRetryableError(e)) throw e;
      if (!isRetryableError(e)) throw e;
      lastError = e;

      if (attempt >= maxRetries) break;

      let wait = attempt < backoff.length ? backoff[attempt]! : backoff.at(-1)!;
      // The provider's own Retry-After wins over our ladder: it knows when its
      // quota resets and we do not.
      if (respectRetryAfter && e instanceof ProviderError && e.retryAfter !== null) {
        wait = e.retryAfter;
      }

      logCallback?.(
        'warning',
        `Retry ${attempt + 1}/${maxRetries} after ${(e as Error).name}: ` +
          `${(e as Error).message} — waiting ${wait}s`,
      );
      await interruptibleSleep(wait, cancelCheck);
    }
  }

  throw lastError;
}
